### ISSUE TITLE
Solar bias correction: add minimum sample gate and fix double-correction

### ISSUE THREAD

**Issue by jackmcintyre** [jackmcintyre/localshift#757]
## Problem Summary

Solar forecast is showing ~44% higher than Solcast predictions due to the solar bias correction system overcompensating with insufficient training data.

### Evidence (from Issue #756)
| Time Slot | LocalShift | Solcast | Discrepancy |
|-----------|------------|---------|-------------|
| 10:00-10:30 | 4.03 kWh | 2.79 kWh | +44% |
| 10:30-11:00 | 4.09 kWh | 2.83 kWh | +44% |
| 11:00-11:30 | 4.30 kWh | 2.99 kWh | +44% |

### Solar Accuracy Metrics (from Issue #756)
| Metric | Value | Interpretation |
|--------|-------|----------------|
| Accuracy | 27.6% | Very poor |
| Bias | -1.76 | Solcast appears to underestimate by 176% |
| MAPE | 72.4% | Mean Absolute Percentage Error |
| Sample count | **4** | ⚠️ Only 4 data points |

### Calculation Chain Causing Inflation
```
Solcast forecast (e.g., 2.79 kWh)
  → Code interprets units correctly as kW average (kWh/hour) ✓
  → Bias correction multiplier = 1.0 - (-1.76) = 2.76x
  → Result: 2.79 * 2.76 = 7.70 kWh (should be ~2.79 kWh)
```

**The real-time solar reading (1.255 kW) matches Solcast prediction (1.274 kW), confirming Solcast is accurate. The issue is purely in the bias correction system.**

---

### Root Cause 1: No Minimum Sample Gate (PRIMARY)

**Location**: `custom_components/localshift/forecast/solar_accuracy.py`

The `SolarAccuracyTracker` applies bias corrections without any minimum sample threshold:

- **`get_bias_correction()`** (line ~488): Returns `1.0 - weighted_bias` even with 1 sample
- **`get_additive_correction()`** (line ~526): Returns clamped offset even with 1 sample
- **`apply_bias_correction()`** (line ~542): Applies combined correction even with 1 sample

**Inconsistency with other subsystems**:
- `ForecastCorrectionProvider` (`forecast/corrections.py`, line 66-73): Requires `MIN_CORRECTION_SAMPLES = 10` before returning correction
- `PatternAnalyzer` (`engine/pattern_analyzer.py`, lines 274-289): Requires `MIN_SAMPLES_FOR_BIAS = 10` AND `MIN_WEEKS_OBSERVED = 2`

The solar accuracy tracker is the ONLY correction system with no gating.

### Root Cause 2: Double Correction

**Location**: `custom_components/localshift/forecast/solar_accuracy.py`, line ~542

```python
def apply_bias_correction(
    self,
    forecast_kwh: float,
    time_of_day: str,
    weather_condition: str,
    season: str | None = None,
) -> float:
    multiplier = 1.0 - self.get_bias_correction(time_of_day, weather_condition, season)
    additive_offset = self.get_additive_correction(time_of_day, weather_condition, season)
    return max(0.0, forecast_kwh * multiplier - additive_offset)
```

**The problem**: Both corrections address the same underlying error:
- `multiplier = 1.0 - bias` accounts for proportional error (e.g., Solcast is 20% high → multiply by 0.8)
- `additive_offset = mean(forecast - actual)` accounts for fixed offset error

If Solcast consistently forecasts 20% high with no fixed offset:
- `bias = 0.20`, `multiplier = 0.80`
- `additive_offset = 0.20 * forecast`
- Result: `forecast * 0.80 - (0.20 * forecast) = 0.60 * forecast` (over-corrects by another 20%)

**This double-counts the same error, causing over-correction.**

### Root Cause 3: Layered Correction Compounding

**Location 1**: `custom_components/localshift/engine/slots.py`, line ~344-347

```python
def _get_solar_kwh(...) -> float:
    solar_kwh = get_solar_for_slot_by_interval(all_solcast, slot_start, interval_minutes)
    return max(0.0, solar_kwh * solar_confidence_factor)  # First correction
```

**Location 2**: `custom_components/localshift/engine/optimizer_facade.py`, line ~82-115

```python
def _apply_bias_correction_to_slots(self, slots: list[Any], weather_condition: str) -> None:
    ...
    slot.solar_kwh = self._solar_accuracy_tracker.apply_bias_correction(
        slot.solar_kwh,  # Already multiplied by solar_confidence_factor
        time_of_day, weather_condition, season,
    )
```

**Order of operations** (from `optimizer_facade.py` line ~180-187):
1. `SlotBuilder.build_slots()` → raw forecast
2. `_record_forecasts_for_slots()` → record for accuracy tracking
3. `_get_solar_kwh()` → applies `solar_confidence_factor` (first correction)
4. `_apply_bias_correction_to_slots()` → applies bias correction multiplier (second correction)

**The problem**: Both systems correct for the same underlying issue (Solcast bias) but apply corrections independently. If Solcast is 20% high:
- `solar_confidence_factor` might learn 0.8x
- `bias_correction` might learn 0.8x
- Combined: 0.8 × 0.8 = 0.64x (over-correction by another 20%)

---

### 1. Add Minimum Sample Gate

**File**: `custom_components/localshift/forecast/solar_accuracy.py`

Add constant near top of file (with other constants like `MAX_PERIOD_RECORDS`, `BIAS_HALF_LIFE_DAYS`):
```python
MIN_SOLAR_CORRECTION_SAMPLES = 20
```

Gate the three correction methods:

**`get_bias_correction()`** (around line 488):
```python
def get_bias_correction(
    self, time_of_day: str, weather_condition: str, season: str | None = None
) -> float:
    # NEW: Require minimum samples before applying correction
    if len(self._records) < MIN_SOLAR_CORRECTION_SAMPLES:
        return 0.0  # No proportional bias correction
    
    weighted_bias = self._compute_context_bias(time_of_day, weather_condition, season)
    return weighted_bias
```

**`get_additive_correction()`** (around line 526):
```python
def get_additive_correction(
    self, time_of_day: str, weather_condition: str, season: str | None = None
) -> float:
    # NEW: Require minimum samples before applying correction
    if len(self._records) < MIN_SOLAR_CORRECTION_SAMPLES:
        return 0.0
    
    weighted_additive = self._compute_context_additive_bias(time_of_day, weather_condition, season)
    return max(-MAX_ADDITIVE_OFFSET_KWH, min(MAX_ADDITIVE_OFFSET_KWH, weighted_additive))
```

**`apply_bias_correction()`** (around line 542):
```python
def apply_bias_correction(
    self,
    forecast_kwh: float,
    time_of_day: str,
    weather_condition: str,
    season: str | None = None,
) -> float:
    # NEW: Require minimum samples before applying any correction
    if len(self._records) < MIN_SOLAR_CORRECTION_SAMPLES:
        return forecast_kwh  # Pass through unchanged
    
    multiplier = 1.0 - self.get_bias_correction(time_of_day, weather_condition, season)
    return max(0.0, forecast_kwh * multiplier)
```

**Add property for external queries**:
```python
@property
def correction_active(self) -> bool:
    """Whether bias correction is active (sufficient samples collected)."""
    return len(self._records) >= MIN_SOLAR_CORRECTION_SAMPLES
```

**Why 20 samples?** With 30-minute recording intervals during daylight (~10-14 samples/day), 20 samples = ~2 days of data. This balances responsiveness with statistical reliability. The issue suggested 50, but 20 is consistent with the approach of "gather evidence quickly, refine over time."

### 2. Remove Additive Correction

**File**: `custom_components/localshift/forecast/solar_accuracy.py`

In `apply_bias_correction()`, remove the additive offset line entirely:

```python
def apply_bias_correction(
    self,
    forecast_kwh: float,
    time_of_day: str,
    weather_condition: str,
    season: str | None = None,
) -> float:
    if len(self._records) < MIN_SOLAR_CORRECTION_SAMPLES:
        return forecast_kwh
    
    multiplier = 1.0 - self.get_bias_correction(time_of_day, weather_condition, season)
    # REMOVED: additive_offset = self.get_additive_correction(...)
    return max(0.0, forecast_kwh * multiplier)
```

**Rationale**: Multiplicative correction scales naturally with forecast magnitude. If Solcast is consistently 20% high, the bias will be ~0.20 and multiplier ~0.80. The additive correction doesn't scale with forecast size and double-counts the proportional error.

**Note**: The `get_additive_correction()` method can remain for diagnostics (returning 0.0 when under threshold), but it won't be used in the final calculation.

### 3. Bias Correction Supersedes solar_confidence_factor

**File**: `custom_components/localshift/engine/optimizer_facade.py`

Modify `_apply_bias_correction_to_slots()` to communicate whether bias correction was applied:

```python
def _apply_bias_correction_to_slots(
    self, slots: list[Any], weather_condition: str
) -> bool:  # NEW: Return whether correction was applied
    if self._solar_accuracy_tracker is None:
        return False
    
    # Check if bias correction will be active (sufficient samples)
    bias_active = self._solar_accuracy_tracker.correction_active
    
    corrected = 0
    for slot in slots:
        slot_dt = datetime.fromisoformat(slot.timestamp_iso)
        time_of_day = self._get_time_of_day(slot_dt)
        season = self._get_season(slot_dt)
        original = slot.solar_kwh
        slot.solar_kwh = self._solar_accuracy_tracker.apply_bias_correction(
            slot.solar_kwh,
            time_of_day,
            weather_condition,
            season,
        )
        if abs(slot.solar_kwh - original) > 0.001:
            corrected += 1
    
    return bias_active  # NEW: Return whether we're using bias correction
```

Then in `run_inline()` (around line 180), pass this info to control `solar_confidence_factor` application:

```python
slots, slot_metadata = slot_builder.build_slots(
    data, data.adaptive_params, now_dt=now_dt
)

weather_condition = getattr(data, "weather_condition", None) or "unknown"
self._record_forecasts_for_slots(slots, weather_condition)
bias_is_active = self._apply_bias_correction_to_slots(slots, weather_condition)

# NEW: If bias correction is active, temporarily override solar_confidence_factor to 1.0
if bias_is_active:
    original_confidence = data.adaptive_params.solar_confidence_factor
    data.adaptive_params.solar_confidence_factor = 1.0

self._apply_cloud_scale_factor_to_slots(slots, data, now_dt)
result = self._planner.plan(inputs)

# NEW: Restore original value if we overrode it
if bias_is_active:
    data.adaptive_params.solar_confidence_factor = original_confidence
```

**Rationale**: Bias correction is more sophisticated (context-aware by time_of_day/weather/season) while `solar_confidence_factor` is a blunt global parameter. When we have sufficient data for context-aware correction, it should supersede the blunt instrument.

### 4. Leave Existing Data As-Is

The 4 unreliable samples with -1.76 bias will naturally be phased out by:
- The 7-day half-life exponential decay in `_compute_context_bias()`
- The accumulation of good data (16 more samples = 20 total before correction activates)

No manual reset needed. The system will "heal itself" as more data accumulates.

---

## Files to Modify

| File | Changes |
|------|---------|
| `custom_components/localshift/forecast/solar_accuracy.py` | Add `MIN_SOLAR_CORRECTION_SAMPLES = 20`, gate 3 correction methods, add `correction_active` property |
| `custom_components/localshift/engine/optimizer_facade.py` | Modify `_apply_bias_correction_to_slots()` to return bool, check in `run_inline()` to bypass `solar_confidence_factor` |

---

## Test Coverage Needed

1. **Unit tests for threshold gating** (`tests/forecast/test_solar_accuracy.py`):
   - `get_bias_correction()` returns 0.0 when under 20 samples
   - `get_additive_correction()` returns 0.0 when under 20 samples
   - `apply_bias_correction()` returns input unchanged when under 20 samples
   - `correction_active` property reflects threshold

2. **Unit tests for multiplicative-only correction**:
   - With sufficient samples, only multiplier is applied (no additive subtraction)

3. **Integration test for supersession**:
   - When `correction_active` is True, `solar_confidence_factor` is temporarily set to 1.0 during planning

---

## Related Patterns in Codebase

| Component | Constant | Value | Behavior When Under Threshold |
|-----------|----------|-------|------------------------------|
| `ForecastCorrectionProvider` | `MIN_CORRECTION_SAMPLES` | 10 | Returns 1.0 (no correction) |
| `PatternAnalyzer` | `MIN_SAMPLES_FOR_BIAS` | 10 | Skips bucket |
| `PatternAnalyzer` | `MIN_WEEKS_OBSERVED` | 2 | Skips entire detection |
| **`SolarAccuracyTracker` (this fix)** | **`MIN_SOLAR_CORRECTION_SAMPLES`** | **20** | **Returns 0.0 / input unchanged** |

---

## References

- Original Bug Report: Issue #756
- Solar Forecast Code: `custom_components/localshift/forecast/solar.py`
- Solar Accuracy Tracker: `custom_components/localshift/forecast/solar_accuracy.py`
- Slot Builder: `custom_components/localshift/engine/slots.py` (line ~344)
- Optimizer Facade: `custom_components/localshift/engine/optimizer_facade.py` (lines ~82-115, ~180-187)
- Pattern to Match: `custom_components/localshift/forecast/corrections.py` (lines 66-73)
